tests/integration/suite/actors/healthz/healthz.go

```diff
@@ -39,22 +39,27 @@ func init() {
 // initerror tests that Daprd will block actor calls until actors have been
 // initialized.
 type initerror struct {
-	daprd        *daprd.Daprd
-	place        *placement.Placement
-	configCalled chan struct{}
-	blockConfig  chan struct{}
+	daprd         *daprd.Daprd
+	place         *placement.Placement
+	configCalled  chan struct{}
+	blockConfig   chan struct{}
+	healthzCalled chan struct{}
 }
 
 func (i *initerror) Setup(t *testing.T) []framework.Option {
 	i.configCalled = make(chan struct{})
 	i.blockConfig = make(chan struct{})
+	i.healthzCalled = make(chan struct{})
 
 	handler := http.NewServeMux()
 	handler.HandleFunc("/dapr/config", func(w http.ResponseWriter, r *http.Request) {
 		close(i.configCalled)
 		<-i.blockConfig
 		w.Write([]byte(`{"entities": ["myactortype"]}`))
 	})
+	handler.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
```

Comment:
@JoshVanL need your help with this please, since you just added this test...

It didn't include the /healthz endpoint (required per actors' contract), so it only worked due to a bug I've fixed in this PR where Dapr assumes an app is healthy before it does a healthcheck.

However, unlike for the TTL test, adding a handler for /healthz here doesn't seem to work.

PTAL

Reply:
@ItalyPaleAle  you are missing a call to close the `healthzCalled` channel inside the app `healthz` HTTP handler, as well as enabling App health checks `daprd.WithAppHealthCheck(true)`.

This looks like it should be a separate integration entirely though.

Reply:
> You are missing a call to close the healthzCalled channel inside the app healthz

Whoops. Still doesn't work however

>  as well as enabling App health checks daprd.WithAppHealthCheck(true). This looks like it should be a separate integration entirely though.

No, this is not for app health checks, but for the health checks that are [part of the actors contract](https://docs.dapr.io/reference/api/actors_api/#health-check). Those should have been added from the beginning in this test.

The reason why this test worked without them is due to a bug this PR is fixing. Currently, Dapr considers an actor host healty by default when it starts, and then it can switch it to un-healthy after healthchecks fail. This means that any actor host that doesn't implement /healthz (in violation of the contract) will work for ~20 seconds before Dapr detects it as unhealthy. This PR fixes that bug too by making Dapr wait for a successful healthcheck first.

Reply:
> No, this is not for app health checks, but for the health checks that are [part of the actors contract](https://docs.dapr.io/reference/api/actors_api/#health-check). Those should have been added from the beginning in this test.

This test is about a call to actors blocking until it is ready without returning an error, we are not supposed to wait for actors to be ready before making a call. This test should continue to pass after changes. Please make a new test.

```
// initerror tests that Daprd will block actor calls until actors have been
// initialized.
```

https://github.com/dapr/dapr/blob/c270ab0ea90563a059b1fdc0f2205f0167c58f8b/tests/integration/suite/actors/healthz/healthz.go#L39

```go
/*
Copyright 2023 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implieh.
See the License for the specific language governing permissions and
limitations under the License.
*/

package healthz

import (
	"context"
	"net"
	"net/http"
	"strconv"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/dapr/dapr/tests/integration/framework"
	"github.com/dapr/dapr/tests/integration/framework/process/daprd"
	prochttp "github.com/dapr/dapr/tests/integration/framework/process/http"
	"github.com/dapr/dapr/tests/integration/framework/process/placement"
	"github.com/dapr/dapr/tests/integration/framework/util"
	"github.com/dapr/dapr/tests/integration/suite"
)

func init() {
	suite.Register(new(healthz))
}

type healthz struct {
	daprd         *daprd.Daprd
	place         *placement.Placement
	healthzCalled chan struct{}
}

func (h *healthz) Setup(t *testing.T) []framework.Option {
	h.healthzCalled = make(chan struct{})

	handler := http.NewServeMux()
	handler.HandleFunc("/dapr/config", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`{"entities": ["myactortype"]}`))
	})
	handler.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
		close(h.healthzCalled)
	})
	handler.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`OK`))
	})

	srv := prochttp.New(t, prochttp.WithHandler(handler))
	h.place = placement.New(t)
	h.daprd = daprd.New(t, daprd.WithResourceFiles(`
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: mystore
spec:
  type: state.in-memory
  version: v1
  metadata:
  - name: actorStateStore
    value: true
`),
		daprd.WithPlacementAddresses("localhost:"+strconv.Itoa(h.place.Port())),
		daprd.WithAppProtocol("http"),
		daprd.WithAppPort(srv.Port()),
		daprd.WithAppHealthCheck(true),
	)

	return []framework.Option{
		framework.WithProcesses(h.place, srv, h.daprd),
	}
}

func (h *healthz) Run(t *testing.T, ctx context.Context) {
	h.place.WaitUntilRunning(t, ctx)

	assert.Eventually(t, func() bool {
		dialer := net.Dialer{Timeout: time.Second}
		net, err := dialer.DialContext(ctx, "tcp", "localhost:"+strconv.Itoa(h.daprd.HTTPPort()))
		if err != nil {
			return false
		}
		net.Close()
		return true
	}, time.Second*5, time.Millisecond*100)

	select {
	case <-h.healthzCalled:
	case <-time.After(time.Second * 5):
		t.Fatal("timed out waiting for healthz call")
	}

	client := util.HTTPClient(t)

	rctx, cancel := context.WithTimeout(ctx, time.Second*2)
	t.Cleanup(cancel)
	daprdURL := "http://localhost:" + strconv.Itoa(h.daprd.HTTPPort()) + "/v1.0/actors/myactortype/myactorid/method/foo"

	req, err := http.NewRequestWithContext(rctx, http.MethodPost, daprdURL, nil)
	require.NoError(t, err)
	resp, err := client.Do(req)
	assert.ErrorIs(t, err, context.DeadlineExceeded)
	if resp != nil && resp.Body != nil {
		assert.NoError(t, resp.Body.Close())
	}
}
```

Reply:
Thanks. This worked, although I still had to add the healthz endpoint to the initerror test, and wait for healthz to be called, before the test worked.

Reply:
@ItalyPaleAle please remove the healthz check from the test- the test is designed specifically to handle the case where actor calls should hang until actors have been initialized. Needing to change the test in order for it to pass signals that this is a breaking change. What would have caused the test to now fail?

With this change- the healthz check takes at least 5 seconds in order to continue. I'd say this is way too much time (especially for integration tests). Scaling to 100s of replicas will result in dozens of minutes of  stalled deployments. Can we not reduce this to, say, 1 or half a second?

> It didn't include the /healthz endpoint (required per actors' contract),

Thinking about this- should we not be disabling actors completely if this condition is not met? (just like we do with needing an actor state store).

Reply:
>  please remove the healthz check from the test- the test is designed specifically to handle the case where actor calls should hang until actors have been initialized. Needing to change the test in order for it to pass signals that this is a breaking change. What would have caused the test to now fail?

No, not a breaking change. As I mentioned, the current behavior is a bug.

The [actors contract](https://docs.dapr.io/reference/api/actors_api/#health-check) requires apps that have an app channel implement /healthz. This is in the contract and it's implemented by all SDKs.

The reason why it wasn't failing before was due to a bug where Dapr would assume the app to be healthy until proven otherwise. That is: when the healthchecks start, the initial state is "healthy", and then it takes 2 failed checks to move the app back to un-healthy (switching from healthy to un-healthy causes placement state to change, so it's a very costly operation for the entire cluster as it causes actors to be rebalanced).

The actor healthchecks [aren't configurable](https://github.com/dapr/dapr/blob/master/pkg/health/health.go#L29-L33) and perform healthchecks every 5s with a timeout of 2s. So, after Dapr starts, if the app isn't healthy it takes up to 14s for Dapr to consider the app unhealthy.

During those 14s, actors were working, and that's why your test was passing, because it completed in less than 14s. But, that's clearly a bug and it's against the contract.

> With this change- the healthz check takes at least 5 seconds in order to continue. I'd say this is way too much time (especially for integration tests). Scaling to 100s of replicas will result in dozens of minutes of stalled deployments. Can we not reduce this to, say, 1 or half a second?

As I mentioned above, the current options aren't configurable. The ticker runs every 5s and has a timeout of 2s, so depending on where the ticker is, it can take a few seconds for the test to pass.

Making these options configurable may not be a terrible idea (I'd probably make them an option in the Configuration CRD), but it's out of the scope of this PR.

> Thinking about this- should we not be disabling actors completely if this condition is not met? (just like we do with needing an actor state store).

That **is** a breaking change.

- Currently, /healthz  for actors is required IF AND ONLY IF the app has a HTTP channel.
- If the app has no app channel, or if the app channel is based on gRPC, then /healthz is not required and the app is considered always healthy
- Using Dapr workflow today doesn't require an app channel, so disabling actors if there's no /healthz would break workflow as it's implemented today.

In the case of Workflow, note that the current E2E test has the same bug as this Integration test, and I [fixed it here](https://github.com/dapr/dapr/pull/6966/files#diff-403da46a67fe1e4bc0cef2f29b5112c22945b9734f346d77fe2f2bd06d162e8eR58). The E2E test deployed the Workflow app with an app channel, even though it wasn't used. So, that made /healthz required. But the WF app doesn't have a server, so tests were now failing. Before this PR, tests were passing due to the bug mentioned above where for the first ~14s the app was considered healthy (and the test completed in less than that - but of course that's not a real-world scenario)

Reply:
OK, thank you for explaining @ItalyPaleAle. I understand now that `healthz` is required but because of a bug the previous behaviour was working.

I do think that the current values are very inappropriate for a sidecar HTTP health check, especially where the app is in most circumstances completely stateless. I am going to create a PR to reduce these values in total, and have more aggressive intervals when in a non-healthy state. We can also look into making these values configurable but this would be a good start.